In [17]:
import pandas as pd

df = pd.read_csv('../data/RMBR4-2_export_test.csv')

df.head()

,Trait,Axis #1,Axis #2,Axis #3,Axis #4,Axis #5,Axis #6,Axis #7,Axis #8,Axis #9,Axis #10,Axis #11,Axis #12,Axis #13,Axis #14,Time
0,current,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,2022-10-17T12:18:23.660Z
1,current,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,2022-10-17T12:18:25.472Z
2,current,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,2022-10-17T12:18:27.348Z
3,current,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,2022-10-17T12:18:29.222Z
4,current,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,2022-10-17T12:18:31.117Z


In [3]:
df.describe()

,Axis #1,Axis #2,Axis #3,Axis #4,Axis #5,Axis #6,Axis #7,Axis #8,Axis #9,Axis #10,Axis #11,Axis #12,Axis #13,Axis #14
count,39672.000000,39672.000000,39672.000000,39672.000000,39672.000000,39672.000000,39672.000000,39672.000000,0.0,0.0,0.0,0.0,0.0,0.0
mean,0.725743,3.613374,2.710336,0.620222,0.954521,0.599427,0.870145,0.102214,NaN,NaN,NaN,NaN,NaN,NaN
std,2.162120,6.879962,5.111901,1.574897,2.100186,1.815498,2.166811,0.423075,NaN,NaN,NaN,NaN,NaN,NaN
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN,NaN,NaN,NaN,NaN,NaN
25%,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN,NaN,NaN,NaN,NaN,NaN
50%,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN,NaN,NaN,NaN,NaN,NaN
75%,0.312710,4.217190,4.586190,0.516190,0.800090,0.361330,0.310030,0.078480,NaN,NaN,NaN,NaN,NaN,NaN
max,23.609300,51.713230,41.855560,15.666300,20.750760,20.931420,8.108480,5.905640,NaN,NaN,NaN,NaN,NaN,NaN


## Data Inspection

The data set contains 39,672 robotic telemetry records and 16 columns.

Axis 9 to 14 do not contain data, however they will be left in the query so that the data remains intact

In [18]:
empty_columns = df.columns[df.isna().all()]

print("Completely empty columns:")
print(empty_columns.tolist())

Completely empty columns:
['Axis #9', 'Axis #10', 'Axis #11', 'Axis #12', 'Axis #13', 'Axis #14']


In [7]:
df["Time"] = pd.to_datetime(df["Time"])

df["Time"].dtype
#df.dtypes

datetime64[us, UTC]

### Streaming Data into Memory

The CSV contains historical robot telemetry. To simulate a live data
stream, the records are read sequentially and added to an in-memory
Pandas DataFrame.

In [8]:
streamed_data = pd.DataFrame(columns=df.columns)

for index, row in df.iterrows():
    streamed_data = pd.concat(
        [streamed_data, row.to_frame().T],
        ignore_index=True
    )

    if index >= 99:
        break

print("Records streamed:", len(streamed_data))

Records streamed: 100


The collected robotic telemetry will be stored in a PostgreSQL relational database hosted in Neon. this in order to be able to modify the data and consult them afterwards

In [9]:
%pip install psycopg2-binary sqlalchemy

Note: you may need to restart the kernel to use updated packages.


In [10]:
from sqlalchemy import create_engine

In [11]:
DATABASE_URL = "postgresql://neondb_owner:npg_s2ZHCKkBPw1J@ep-steep-voice-b4mrhhuh-pooler.c-6.us-east-2.aws.neon.tech/neondb?sslmode=require&channel_binding=require"

engine = create_engine(DATABASE_URL)

with engine.connect() as connection:
    print("Database connection successful!")

Database connection successful!


Connecting to NEON uploading the data from CSV to Postgres

In [12]:
df.to_sql(
    "robot_telemetry",
    engine,
    if_exists="replace",
    index=False,
    chunksize=1000
)

print("Data successfully stored in PostgreSQL!")

Data successfully stored in PostgreSQL!


In [13]:
from sqlalchemy import text

with engine.connect() as connection:
    result = connection.execute(
        text("SELECT COUNT(*) FROM robot_telemetry")
    )
    
    print("Records in database:", result.scalar())

Records in database: 39672


In [16]:
from sqlalchemy import text

with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT *
            FROM robot_telemetry
            LIMIT 5
        """)
    )

    for row in result:
        print(row)

('current', 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, datetime.datetime(2022, 10, 17, 12, 18, 23, 660000, tzinfo=datetime.timezone.utc))
('current', 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, datetime.datetime(2022, 10, 17, 12, 18, 25, 472000, tzinfo=datetime.timezone.utc))
('current', 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, datetime.datetime(2022, 10, 17, 12, 18, 27, 348000, tzinfo=datetime.timezone.utc))
('current', 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, datetime.datetime(2022, 10, 17, 12, 18, 29, 222000, tzinfo=datetime.timezone.utc))
('current', 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, None, None, None, None, None, None, datetime.datetime(2022, 10, 17, 12, 18, 31, 117000, tzinfo=datetime.timezone.utc))
